### DISCUSSION TITLE
Unifying the `-offline` and `-online` variants back into one version

### DISCUSSION THREAD

**Discussion by moneytoo** [moneytoo/Player#119] [General]
I'm unifying the `-offline` and `-online` variants back into one version that will include all the features, quirks etc. I intended to have at least one _small, clean and strict_ variant, but the ideals clash with real life usability.

* **Internet permission** vs streaming support: Theoretically it is possible to access files on network using DocumentsProvider intermediate apps so an app like a video player doesn't need such permission. There are some such apps (for protocols like SFTP, WebDAV, SMB) but none fully working and capable streaming files. Even apps for paid cloud storage work only with progressive download. So in real life, one need the internet permission mostly for access localhost, when file managers proxy files on network.

* **File system access**, TV boxes: Android TV devices do not provide Storage Access Framework and you have to use the "legacy" access. Except some devices actually offer that and some may offer just some non working stubs ([link](https://stackoverflow.com/questions/64082063/on-some-devices-action-open-document-tree-doesnt-show-the-chooser#comment113318551_64082063)). It's a mess and it requires hacks/quirks.

* **APK file size**: Right now the apk file size inflates to over 50 MB. This is primarily because of FFmpeg (ffprobe) dependency for detecting video frame rate. I'm planning on migrating to custom mediainfo library that should be significantly smaller. But also please note that the FFmpeg library in the APK for 5 different architectures so users installing the app from Google Play will get much smaller installation.

Theoretically I could provide variants offline/online, with/without extra ffmpeg audio decoders, with/without frame rate detection, with/without [subtitle encoding detection library](https://github.com/moneytoo/Player/issues/76), for arm/arm64/x86/x86_64/universal. But that would be 2\*2\*2\*2\*5... 80 different APKs.

So it is what it is and that's that.

**Comment by roblav96** [comment]
@moneytoo regarding the frame rate, take a look this PR by koying:

[CHG: Be smarter when switching reso](https://github.com/jellyfin/jellyfin-androidtv/pull/897)

**Reply by roblav96** [reply]
also instead of using FFmpeg, use the [MediaExtractor](https://developer.android.com/reference/android/media/MediaExtractor) class and extract the frame rate with [MediaFormat#KEY_FRAME_RATE](https://developer.android.com/reference/android/media/MediaFormat#KEY_FRAME_RATE)

**Reply by moneytoo** [reply]
> also instead of using FFmpeg, use the [MediaExtractor](https://developer.android.com/reference/android/media/MediaExtractor) class and extract the frame rate with [MediaFormat#KEY_FRAME_RATE](https://developer.android.com/reference/android/media/MediaFormat#KEY_FRAME_RATE)

I tried that before but it doesn't work with all the formats (MP4 is ok, but not Matroska).
